modules/connectors/ninja_van/karrio/providers/ninja_van/rate.py

```diff
@@ -0,0 +1,58 @@
+
+import typing
+import karrio.lib as lib
+import karrio.core.units as units
+import karrio.core.models as models
+import karrio.providers.ninja_van.error as error
+import karrio.providers.ninja_van.utils as provider_utils
+import karrio.providers.ninja_van.units as provider_units
+
+
+def parse_rate_response(
+    _response: lib.Deserializable[dict],
+    settings: provider_utils.Settings,
+) -> typing.Tuple[typing.List[models.RateDetails], typing.List[models.Message]]:
+    response = _response.deserialize()
+
+    messages = error.parse_error_response(response, settings)
+    rates = [_extract_details(rate, settings) for rate in response]
+
+    return rates, messages
+
+
+def _extract_details(
+    data: dict,
+    settings: provider_utils.Settings,
+) -> models.RateDetails:
+    rate = None  # parse carrier rate type
+
+    return models.RateDetails(
+        carrier_id=settings.carrier_id,
+        carrier_name=settings.carrier_name,
+        service="",  # extract service from rate
+        total_charge=lib.to_money(0.0),  # extract the rate total rate cost
+        currency="",  # extract the rate pricing currency
+        transit_days=0,  # extract the rate transit days
+        meta=dict(
+            service_name="",  # extract the rate service human readable name
+        ),
+    )
+
+
+def rate_request(
+    payload: models.RateRequest,
+    settings: provider_utils.Settings,
+) -> lib.Serializable:
+    shipper = lib.to_address(payload.shipper)
+    recipient = lib.to_address(payload.recipient)
+    packages = lib.to_packages(payload.parcels)
+    services = lib.to_services(payload.services, provider_units.ShippingService)
+    options = lib.to_shipping_options(
+        payload.options,
+        package_options=packages.options,
+    )
+
+    # map data to convert karrio model to ninja_van specific type
+    request = None
```

Comment:
This needs to be implemented as well. Once you have the schema files, this is where you use them to create the request for the implementation.


Here is another good example from Sendle 
https://github.com/karrioapi/karrio/blob/12792cff650d63817206aa86298e6b8201f3dd6b/modules/connectors/sendle/karrio/providers/sendle/rate.py#L97-L119

Notice the types generated by the schema are imported at the top

https://github.com/karrioapi/karrio/blob/12792cff650d63817206aa86298e6b8201f3dd6b/modules/connectors/sendle/karrio/providers/sendle/rate.py#L1-L2

The request is pretty much a mirror of the schema file 
https://github.com/karrioapi/karrio/blob/12792cff650d63817206aa86298e6b8201f3dd6b/modules/connectors/sendle/schemas/product_request.json#L1-L20